### PR TITLE
Feature/upload multiple files at once to bytes

### PR DESCRIPTION
**Pull request by Donnype**
### Changes

This makes the raw endpoint accept multiple files by default. Note that I think that utilizing the new functionality belongs in a subsequent PR, so there is no functionality added here and everything is backward compatible at the BytesClient level at least, mostly to keep PR's small and focused.

### Issue link

Closes https://github.com/minvws/nl-kat-coordination/issues/3113

### Demo

-

### QA notes

These changes do not add functionality but affect Bytes, Boefjes and Rocky. Please verify that the boefjes still run properly, we can download raw files from Rocky, and we can still upload raw files from Rocky!

---

### Code Checklist

- [x] All the commits in this PR are properly PGP-signed and verified.
- [x] This PR only contains functionality relevant to the issue.
- [x] I have written unit tests for the changes or fixes I made.
- [x] I have checked the documentation and made changes where necessary.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.
- [x] For any non-trivial functionality, I have added integration and/or end-to-end tests.

---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: boefjes/boefjes/clients/bytes_client.py [hunk 1]
```diff
@@ -99,17 +100,25 @@ def get_normalizer_meta(self, normalizer_meta_id: uuid.UUID) -> NormalizerMeta:
 
     @retry_with_login
     def save_raw(self, boefje_meta_id: str, raw: str | bytes, mime_types: Set[str] = frozenset()) -> UUID:
-        headers = {"content-type": "application/octet-stream"}
-        headers.update(self.headers)
+        file_name = "raw"  # The name provides a key for all ids returned, so this is arbitrary as we only upload 1 file
```
Comment (underdarknl) [inline boefjes/boefjes/clients/bytes_client.py:103]: Theres a ticket somewhere for naming raw files, as naming them seems logical as a feature and useful for the user in some situations where one job has multiple raw files with the same mime-types..

Comment (Donnype) [reply boefjes/boefjes/clients/bytes_client.py:103]: Oki, let me know if you can find it and I'll take a look at that one as well

**Comment by stephanie0x00** [comment]
During QA I ran into the following TypeError in the onboarding: 

image


```
rocky-1  | {"event": "{\"code\": 500, \"request\": \"POST /en/onboarding/aa/step/setup-scan/URL/?report_type=dns-report\", \"exc_info\": true, \"event\": \"request_failed\", \"user_id\": 1, \"organization_code\": \"aa\", \"ip\": \"172.30.0.1\", \"request_id\": \"d246bc51-a5f6-49cf-9afe-701934dc8fdd\", \"level\": \"error\", \"timestamp\": \"2024-09-09T13:03:39.295245\"}", "exc_info": ["<class 'TypeError'>", "TypeError('Object of type set is not JSON serializable')", "<traceback object at 0x78795f514880>"]}
rocky-1  | Internal Server Error: /en/onboarding/aa/step/setup-scan/URL/
rocky-1  | Traceback (most recent call last):
rocky-1  |   File "/app/rocky/octopoes/connector/octopoes.py", line 50, in _verify_response
rocky-1  |     response.raise_for_status()
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_models.py", line 761, in raise_for_status
rocky-1  |     raise HTTPStatusError(message, request=request, response=self)
rocky-1  | httpx.HTTPStatusError: Client error '404 Not Found' for url 'http://octopoes_api/aa/object?reference=URL%7Cinternet%7Chttps%3A%2F%2Fmispo.es%2F&valid_time=2024-09-09%2013%3A03%3A38.803831%2B00%3A00'
rocky-1  | For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404
rocky-1  | 
rocky-1  | During handling of the above exception, another exception occurred:
rocky-1  | 
rocky-1  | Traceback (most recent call last):
rocky-1  |   File "/app/rocky/tools/ooi_helpers.py", line 247, in get_or_create_ooi
rocky-1  |     return api_connector.get(ooi.reference, observed_at), False
rocky-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/app/rocky/octopoes/connector/octopoes.py", line 96, in get
rocky-1  |     res = self.session.get(
rocky-1  |           ^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 1054, in get
rocky-1  |     return self.request(
rocky-1  |            ^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 827, in request
rocky-1  |     return self.send(request, auth=auth, follow_redirects=follow_redirects)
rocky-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 914, in send
rocky-1  |     response = self._send_handling_auth(
rocky-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 942, in _send_handling_auth
rocky-1  |     response = self._send_handling_redirects(
rocky-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 999, in _send_handling_redirects
rocky-1  |     raise exc
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 982, in _send_handling_redirects
rocky-1  |     hook(response)
rocky-1  |   File "/app/rocky/octopoes/connector/octopoes.py", line 54, in _verify_response
rocky-1  |     raise ObjectNotFoundException(data["detail"])
rocky-1  | octopoes.models.exception.ObjectNotFoundException: URL|internet|https://mispo.es/
rocky-1  | 
rocky-1  | During handling of the above exception, another exception occurred:
rocky-1  | 
rocky-1  | Traceback (most recent call last):
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/core/handlers/exception.py", line 55, in inner
rocky-1  |     response = get_response(request)
rocky-1  |                ^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/core/handlers/base.py", line 197, in _get_response
rocky-1  |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
rocky-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 104, in view
rocky-1  |     return self.dispatch(request, *args, **kwargs)
rocky-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/contrib/auth/mixins.py", line 109, in dispatch
rocky-1  |     return super().dispatch(request, *args, **kwargs)
rocky-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/views/generic/base.py", line 143, in dispatch
rocky-1  |     return handler(request, *args, **kwargs)
rocky-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/django/views/generic/edit.py", line 151, in post
rocky-1  |     return self.form_valid(form)
rocky-1  |            ^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/app/rocky/onboarding/views.py", line 149, in form_valid
rocky-1  |     ooi = self.get_or_create_url_object(cleaned_url)
rocky-1  |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/app/rocky/onboarding/views.py", line 144, in get_or_create_url_object
rocky-1  |     url_ooi, _ = get_or_create_ooi(self.octopoes_api_connector, self.bytes_client, url, observed_at)
rocky-1  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/app/rocky/tools/ooi_helpers.py", line 249, in get_or_create_ooi
rocky-1  |     create_ooi(api_connector, bytes_client, ooi, observed_at, end_valid_time)
rocky-1  |   File "/app/rocky/tools/ooi_helpers.py", line 262, in create_ooi
rocky-1  |     bytes_client.add_manual_proof(task_id, BytesClient.raw_from_declarations([declaration]))
rocky-1  |   File "/app/rocky/rocky/bytes_client.py", line 58, in add_manual_proof
rocky-1  |     raw_id = self._save_raw(boefje_meta.id, raw, all_mime_types)
rocky-1  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/app/rocky/rocky/bytes_client.py", line 119, in _save_raw
rocky-1  |     response = self.session.post(
rocky-1  |                ^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 1145, in post
rocky-1  |     return self.request(
rocky-1  |            ^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 814, in request
rocky-1  |     request = self.build_request(
rocky-1  |               ^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_client.py", line 357, in build_request
rocky-1  |     return Request(
rocky-1  |            ^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_models.py", line 340, in __init__
rocky-1  |     headers, stream = encode_request(
rocky-1  |                       ^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_content.py", line 212, in encode_request
rocky-1  |     return encode_json(json)
rocky-1  |            ^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/site-packages/httpx/_content.py", line 175, in encode_json
rocky-1  |     body = json_dumps(json).encode("utf-8")
rocky-1  |            ^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/json/__init__.py", line 231, in dumps
rocky-1  |     return _default_encoder.encode(obj)
rocky-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/json/encoder.py", line 200, in encode
rocky-1  |     chunks = self.iterencode(o, _one_shot=True)
rocky-1  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/json/encoder.py", line 258, in iterencode
rocky-1  |     return _iterencode(o, 0)
rocky-1  |            ^^^^^^^^^^^^^^^^^
rocky-1  |   File "/usr/local/lib/python3.11/json/encoder.py", line 180, in default
rocky-1  |     raise TypeError(f'Object of type {o.__class__.__name__} '
rocky-1  | TypeError: Object of type set is not JSON serializable
rocky-1  | {"event": "Internal Server Error: /en/onboarding/aa/step/setup-scan/URL/", "exc_info": ["<class 'TypeError'>", "TypeError('Object of type set is not JSON serializable')", "<traceback object at 0x78795e1af800>"]}

```

File: bytes/bytes/api/router.py [hunk 2]
```diff
@@ -148,42 +148,59 @@ def get_normalizer_meta(
...
-                mime_types=raw_data.mime_types,
-            ),
-        )
-        event_manager.publish(event)
-    except Exception as error:
-        logger.exception("Error saving raw data")
-        raise HTTPException(status_code=500, detail="Could not save raw data") from error
-
-    return RawResponse(status="success", message="Raw data saved", id=raw_id)
+) -> JSONResponse:
```
Comment (ammar92) [inline bytes/bytes/api/router.py:155]: I suggest using a more specific type for this, as it describes the function better. FastAPI takes care of the rest, producing a correct OpenAPI specification and wrapping it in the right response object

```suggestion
) -> dict[str, str]:
```

File: bytes/bytes/api/router.py [hunk 3]
```diff
@@ -148,42 +148,59 @@ def get_normalizer_meta(
...
+                ),
+            )
+            event_manager.publish(event)
+        except Exception as error:
+            logger.exception("Error saving raw data")
+            raise HTTPException(status_code=500, detail="Could not save raw data") from error
+
+        all_parsed_mime_types.append(parsed_mime_types)
+
+    return JSONResponse(raw_ids)
```
Comment (ammar92) [inline bytes/bytes/api/router.py:203]: See remark above

```suggestion
    return raw_ids
```

File: bytes/bytes/models.py [hunk 4]
```diff
@@ -38,6 +38,9 @@ def _validate_timezone_aware_datetime(value: datetime) -> datetime:
 class MimeType(BaseModel):
     value: str
 
+    def __hash__(self):
+        return hash(self.value)
```
Comment (ammar92) [inline bytes/bytes/models.py:42]: 👍

File: bytes/bytes/api/router.py [hunk 5]
```diff
@@ -148,42 +148,59 @@ def get_normalizer_meta(
...
+
+        if parsed_mime_types in mime_types_by_id.values():
+            # Set the id for this file using the precomputed dict that maps existing primary keys to the mime-type set.
+            raw_ids[raw.name] = str(
+                list(mime_types_by_id.keys())[list(mime_types_by_id.values()).index(parsed_mime_types)]
+            )
+            continue
+
+        if parsed_mime_types in all_parsed_mime_types:
+            raise HTTPException(status_code=400, detail="Content types do not define unique sets of mime types.")
```
Comment (ammar92) [inline bytes/bytes/api/router.py:176]: Please use the status codes constants defined in `httpx.codes`

```suggestion
            raise HTTPException(status_code=codes.NOT_FOUND, detail="Content types do not define unique sets of mime types.")
```

File: bytes/bytes/api/router.py [hunk 6]
```diff
@@ -148,42 +148,59 @@ def get_normalizer_meta(
...
+                raw_data=RawDataMeta(
+                    id=raw_id,
+                    boefje_meta=raw_data.boefje_meta,
+                    mime_types=raw_data.mime_types,
+                ),
+            )
+            event_manager.publish(event)
+        except Exception as error:
+            logger.exception("Error saving raw data")
+            raise HTTPException(status_code=500, detail="Could not save raw data") from error
```
Comment (ammar92) [inline bytes/bytes/api/router.py:199]: See earlier remark about codes

**Review by ammar92 — COMMENTED** [review]
Nice work! I've very little remarks that might improve it even more. There's still a little issue in the integration tests, but seems like a fixture mismatch. Should be fixable

**Comment by stephanie0x00** [comment]
## Checklist for QA:

- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
  - [x] I have managed to pass the onboarding flow
  - [x] Objects and Findings are created properly
  - [x] Tasks are created and completed properly
- [x] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.
- [x] I checked the logs for errors and/or warnings and made issues where necessary

### What works:
Looks good. Can generate and open reports, tasks and findings are created. Raw files can be uploaded (dnsrecords). 

### What doesn't work:
n/a

### Bug or feature?:
n/a

**Pull request merged**